src/nsrt_learning/strips_learning/clustering_learner.py

```diff
@@ -129,8 +142,168 @@ class ClusterAndSearchSTRIPSLearner(ClusteringSTRIPSLearner):
+            # Construct negative data by merging the datastores of all
+            # other PNADs that have the same option.
+            negative_data = []
+            for j, other_pnad in enumerate(pnads):
+                if i == j:
+                    continue
+                if pnad.option_spec[0] != other_pnad.option_spec[0]:
+                    continue
+                negative_data.extend(other_pnad.datastore)
+            all_preconditions = self._run_search(pnad, positive_data,
+                                                 negative_data)
+            # The datastores of the new PNADs could need to be changed, so
+            # we initialize them as empty, then recompute them at the end.
+            for j, preconditions in enumerate(all_preconditions):
+                new_pnads.append(
+                    PartialNSRTAndDatastore(
+                        pnad.op.copy_with(name=f"{pnad.op.name}-{j}",
+                                          preconditions=preconditions), [],
+                        pnad.option_spec))
+        self._recompute_datastores_from_segments(new_pnads)
+        return new_pnads
+
+    def _run_search(self, pnad: PartialNSRTAndDatastore,
+                    positive_data: Datastore,
+                    negative_data: Datastore) -> Set[FrozenSet[LiftedAtom]]:
+        """Run outer-level search to find a set of precondition sets.
+
+        Each precondition set will produce one operator.
+        """
+        all_preconditions = set()
+        # We'll remove positives as they get covered.
+        remaining_positives = list(positive_data)
+        while remaining_positives:
+            new_preconditions = self._run_single_search(
+                pnad, remaining_positives, negative_data)
+            # Update the remaining positives.
+            new_remaining_positives = []
+            for seg, var_to_obj in remaining_positives:
+                ground_pre = {a.ground(var_to_obj) for a in new_preconditions}
+                if not ground_pre.issubset(seg.init_atoms):
+                    # If the preconditions ground with this substitution don't
+                    # hold in this segment's init_atoms, this segment has yet
+                    # to be covered, so we keep it in the positives.
+                    new_remaining_positives.append((seg, var_to_obj))
+                else:
+                    # Otherwise, we can move this segment to negative_data,
+                    # for any future preconditions that get learned.
+                    negative_data.append((seg, var_to_obj))
+            assert len(new_remaining_positives) < len(remaining_positives)
+            remaining_positives = new_remaining_positives
+            # Update the set to be returned.
+            assert new_preconditions not in all_preconditions
+            all_preconditions.add(new_preconditions)
+        return all_preconditions
+
+    def _run_single_search(self, pnad: PartialNSRTAndDatastore,
```

Comment:
this looks like GBFS. can we use `utils.run_gbfs`?

Reply:
good suggestion, fixing that now